File: client/src/types/lead.ts
export const STAGE = {
  PROSPECT: "PROSPECT",
  ENGAGED: "ENGAGED",
  NEGOTIATION: "NEGOTIATION",
  COMMITTED: "COMMITTED",
  CLOSED: "CLOSED",
} as const;

export type StageType = (typeof STAGE)[keyof typeof STAGE];

export const STAGE_LEVELS: Record<StageType, number> = {
  PROSPECT: 1,
  ENGAGED: 2,
  NEGOTIATION: 3,
  COMMITTED: 4,
  CLOSED: 5,
};

export interface ILead {
  id?: number;
  name: string;
  email: string;
  company_id: number | null;
  company_name: string;
  stage: StageType;
  last_contacted_at: string | null;
  engaged: boolean;
}

File: server/app/schemas/lead.py
from pydantic import BaseModel, EmailStr
from datetime import datetime
from enum import Enum
from typing import List, Optional


class Stage(str, Enum):
    PROSPECT = "PROSPECT"
    ENGAGED = "ENGAGED"
    NEGOTIATION = "NEGOTIATION"
    COMMITTED = "COMMITTED"
    CLOSED = "CLOSED"


class LeadBase(BaseModel):
    name: str
    email: EmailStr
    company_id: int
    engaged: bool = False
    stage: Stage
    last_contacted_at: Optional[datetime] = None


class LeadCreate(BaseModel):
    name: str
    email: EmailStr
    engaged: bool = False
    stage: Stage
    last_contacted_at: Optional[datetime] = None
    company_name: Optional[str] = None
    company_id: Optional[int] = None


class LeadUpdate(BaseModel):
    name: Optional[str] = None
    email: Optional[EmailStr] = None
    company_name: Optional[str] = None
    company_id: Optional[int] = None
    engaged: Optional[bool] = None
    stage: Optional[Stage] = None
    last_contacted_at: Optional[datetime] = None


class BulkDeleteRequest(BaseModel):
    ids: List[int]


class LeadResponse(LeadBase):
    id: int
    company_name: str

    class Config:
        from_attributes = True  # Allows ORM conversion
